dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramObjectBundleHook.java

```diff
@@ -29,43 +29,81 @@
  */
 
 import org.hisp.dhis.common.IdentifiableObject;
+import org.hisp.dhis.dxf2.events.event.AbstractEventService;
+import org.hisp.dhis.dxf2.importsummary.ImportStatus;
+import org.hisp.dhis.dxf2.importsummary.ImportSummary;
 import org.hisp.dhis.dxf2.metadata.objectbundle.ObjectBundle;
-import org.hisp.dhis.program.Program;
-import org.hisp.dhis.program.ProgramStage;
-import org.hisp.dhis.program.ProgramType;
+import org.hisp.dhis.feedback.ErrorCode;
+import org.hisp.dhis.feedback.ErrorReport;
+import org.hisp.dhis.program.*;
 import org.hisp.dhis.security.acl.AccessStringHelper;
+import org.hisp.dhis.user.User;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
 
 /**
  * @author Morten Olav Hansen <mortenoh@gmail.com>
  */
 public class ProgramObjectBundleHook extends AbstractObjectBundleHook
 {
+    private ProgramInstanceService programInstanceService;
+
+    @Autowired
+    public ProgramObjectBundleHook( ProgramInstanceService programInstanceService )
+    {
+        this.programInstanceService = programInstanceService;
+    }
+
     @Override
     public void postCreate( IdentifiableObject object, ObjectBundle bundle )
     {
-        if ( !Program.class.isInstance( object ) )
+        if ( !isProgram( object ) )
         {
             return;
         }
 
         syncSharingForEventProgram( (Program) object );
+
+        addProgramInstance( (Program) object );
     }
 
     @Override
     public void postUpdate( IdentifiableObject object, ObjectBundle bundle )
     {
-        if ( !Program.class.isInstance( object ) )
+        if ( !isProgram( object ) )
         {
             return;
         }
 
         syncSharingForEventProgram( (Program) object );
     }
 
+    @Override
+    public <T extends IdentifiableObject> List<ErrorReport> validate( T object, ObjectBundle bundle )
+    {
+        List<ErrorReport> errors = new ArrayList<>();
+
+        if ( !(object instanceof Program) )
+        {
+            return errors;
+        }
+
+        Program program = (Program) object;
+
+        if ( program.getId() != 0 && getProgramInstancesCount( program ) > 1 )
+        {
+            errors.add( new ErrorReport( Program.class, ErrorCode.E6000, program.getName() ) );
+        }
+        
+        return errors;
+    }
+
     private void syncSharingForEventProgram( Program program )
     {
-        if ( ProgramType.WITH_REGISTRATION == program.getProgramType()
-            || program.getProgramStages().isEmpty() )
+        if ( ProgramType.WITH_REGISTRATION == program.getProgramType() || program.getProgramStages().isEmpty() )
```

Comment:
Can you use `WITHOUT_REGISTRATION` instead? we don't have any plans to add more to this enum, but this would make it a bit more future proof

Reply:
Sorry Morten, I'm not sure I understand this comment. Do you mean that the code should become:

    if ( ProgramType.WITHOUT_REGISTRATION != program.getProgramType() ||    program.getProgramStages().isEmpty() )
        {
            return;
        }

Reply:
Yeah, it's not really important, just saying we might potentially have other types with registration in the future, so its more future proof to only test against what we want, which would be WITHOUT_REGISTRATION.